File: apps/tauri/src-tauri/src/commands/schema.rs
//! Schema commands
//!
//! Handles schema snapshots and comparison operations

use serde::{Deserialize, Serialize};
use serde_json::Value;
use tauri_plugin_store::StoreExt;
use uuid::Uuid;
use chrono::Utc;

const SCHEMA_STORE_FILE: &str = "schema-store.json";
const SNAPSHOTS_KEY: &str = "snapshots";

/// Schema snapshot structure
#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SchemaSnapshot {
    pub id: String,
    pub name: String,
    pub description: Option<String>,
    pub connection_path: String,
    pub schemas: Value,
    pub created_at: String,
}

/// Request to get a specific snapshot
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GetSnapshotRequest {
    pub snapshot_id: String,
}

/// Request to save a snapshot
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SaveSnapshotRequest {
    pub name: String,
    pub description: Option<String>,
    pub connection_path: Option<String>,
    pub schema: Value,
}

/// Request to delete a snapshot
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct DeleteSnapshotRequest {
    pub snapshot_id: String,
}

/// Request to compare snapshots
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompareSnapshotsRequest {
    pub source_snapshot_id: Option<String>,
    pub target_snapshot_id: Option<String>,
    pub snapshot_id1: Option<String>,
    pub snapshot_id2: Option<String>,
}

/// Request to compare connections
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompareConnectionsRequest {
    pub source_connection_id: Option<String>,
    pub target_connection_id: Option<String>,
    pub connection_id1: Option<String>,
    pub connection_id2: Option<String>,
}

/// Request to compare connection to snapshot
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompareConnectionToSnapshotRequest {
    pub connection_id: String,
    pub snapshot_id: String,
}

/// Request to compare tables
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct CompareTablesRequest {
    pub source_connection_id: Option<String>,
    pub target_connection_id: Option<String>,
    pub connection_id1: Option<String>,
    pub connection_id2: Option<String>,
    pub source_table: Option<String>,
    pub target_table: Option<String>,
    pub table1: Option<String>,
    pub table2: Option<String>,
    pub source_schema: Option<String>,
    pub target_schema: Option<String>,
    pub schema1: Option<String>,
    pub schema2: Option<String>,
    pub primary_keys: Option<Vec<String>>,
}

/// Request to generate migration SQL
#[derive(Debug, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct GenerateMigrationSqlRequest {
    pub comparison: Value,
    pub options: Option<Value>,
}

/// Generic response
#[derive(Debug, Serialize)]
pub struct SchemaResponse {
    pub success: bool,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub snapshots: Option<Vec<SchemaSnapshot>>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub snapshot: Option<SchemaSnapshot>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub comparison: Option<Value>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub sql: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub error: Option<String>,
}

fn get_snapshots(app: &tauri::AppHandle) -> Vec<SchemaSnapshot> {
    let store = match app.store(SCHEMA_STORE_FILE) {
        Ok(s) => s,
        Err(_) => return vec![],
    };

    store
        .get(SNAPSHOTS_KEY)
        .and_then(|v| serde_json::from_value(v).ok())
        .unwrap_or_default()
}

fn save_snapshots(app: &tauri::AppHandle, snapshots: &[SchemaSnapshot]) -> Result<(), String> {
    let store = app.store(SCHEMA_STORE_FILE).map_err(|e| e.to_string())?;
    store.set(SNAPSHOTS_KEY, serde_json::to_value(snapshots).map_err(|e| e.to_string())?);
    store.save().map_err(|e| e.to_string())
}

/// Get all schema snapshots
#[tauri::command]
pub async fn schema_get_snapshots(app: tauri::AppHandle) -> SchemaResponse {
    let snapshots = get_snapshots(&app);
    SchemaResponse {
        success: true,
        snapshots: Some(snapshots),
        snapshot: None,
        comparison: None,
        sql: None,
        error: None,
    }
}

/// Get a specific schema snapshot
#[tauri::command]
pub async fn schema_get_snapshot(
    app: tauri::AppHandle,
    request: GetSnapshotRequest,
) -> SchemaResponse {
    let snapshots = get_snapshots(&app);
    let snapshot = snapshots.into_iter().find(|s| s.id == request.snapshot_id);

    SchemaResponse {
        success: snapshot.is_some(),
        snapshots: None,
        snapshot,
        comparison: None,
        sql: None,
        error: None,
    }
}

/// Save a schema snapshot
#[tauri::command]
pub async fn schema_save_snapshot(
    app: tauri::AppHandle,
    request: SaveSnapshotRequest,
) -> SchemaResponse {
    let snapshot = SchemaSnapshot {
        id: Uuid::new_v4().to_string(),
        name: request.name,
        description: request.description,
        connection_path: request.connection_path.unwrap_or_default(),
        schemas: request.schema,
        created_at: Utc::now().to_rfc3339(),
    };

    let mut snapshots = get_snapshots(&app);
    snapshots.push(snapshot.clone());

    if let Err(e) = save_snapshots(&app, &snapshots) {
        return SchemaResponse {
            success: false,
            snapshots: None,
            snapshot: None,
            comparison: None,
            sql: None,
            error: Some(e),
        };
    }

    SchemaResponse {
        success: true,
        snapshots: None,
        snapshot: Some(snapshot),
        comparison: None,
        sql: None,
        error: None,
    }
}

/// Delete a schema snapshot
#[tauri::command]
pub async fn schema_delete_snapshot(
    app: tauri::AppHandle,
    request: DeleteSnapshotRequest,
) -> SchemaResponse {
    let mut snapshots = get_snapshots(&app);
    snapshots.retain(|s| s.id != request.snapshot_id);

    if let Err(e) = save_snapshots(&app, &snapshots) {
        return SchemaResponse {
            success: false,
            snapshots: None,
            snapshot: None,
            comparison: None,
            sql: None,
            error: Some(e),
        };
    }

    SchemaResponse {
        success: true,
        snapshots: None,
        snapshot: None,
        comparison: None,
        sql: None,
        error: None,
    }
}

/// Compare two schema snapshots
#[tauri::command]
pub async fn schema_compare_snapshots(
    app: tauri::AppHandle,
    request: CompareSnapshotsRequest,
) -> SchemaResponse {
    let snapshots = get_snapshots(&app);
    
    let id1 = request.snapshot_id1.or(request.source_snapshot_id).unwrap_or_default();
    let id2 = request.snapshot_id2.or(request.target_snapshot_id).unwrap_or_default();
    
    let snapshot1 = snapshots.iter().find(|s| s.id == id1);
    let snapshot2 = snapshots.iter().find(|s| s.id == id2);

    if snapshot1.is_none() || snapshot2.is_none() {
        return SchemaResponse {
            success: false,
            snapshots: None,
            snapshot: None,
            comparison: None,
            sql: None,
            error: Some("One or both snapshots not found".to_string()),
        };
    }

    // For now, return a placeholder comparison
    // Full comparison logic would be implemented in a dedicated service
    let comparison = serde_json::json!({
        "source": snapshot1.unwrap().name,
        "target": snapshot2.unwrap().name,
        "differences": [],
        "additions": [],
        "deletions": [],
        "modifications": []
    });

    SchemaResponse {
        success: true,
        snapshots: None,
        snapshot: None,
        comparison: Some(comparison),
        sql: None,
        error: None,
    }
}

/// Compare two database connections
#[tauri::command]
pub async fn schema_compare_connections(
    _request: CompareConnectionsRequest,
) -> SchemaResponse {
    // Placeholder - would need database manager access
    SchemaResponse {
        success: true,
        snapshots: None,
        snapshot: None,
        comparison: Some(serde_json::json!({
            "differences": [],
            "additions": [],
            "deletions": [],
            "modifications": []
        })),
        sql: None,
        error: None,
    }
}

/// Compare a connection to a snapshot
#[tauri::command]
pub async fn schema_compare_connection_to_snapshot(
    app: tauri::AppHandle,
    request: CompareConnectionToSnapshotRequest,
) -> SchemaResponse {
    let snapshots = get_snapshots(&app);
    let snapshot = snapshots.iter().find(|s| s.id == request.snapshot_id);

    if snapshot.is_none() {
        return SchemaResponse {
            success: false,
            snapshots: None,
            snapshot: None,
            comparison: None,
            sql: None,
            error: Some("Snapshot not found".to_string()),
        };
    }

    // Placeholder - would need database manager access
    SchemaResponse {
        success: true,
        snapshots: None,
        snapshot: None,
        comparison: Some(serde_json::json!({
            "differences": [],
            "additions": [],
            "deletions": [],
            "modifications": []
        })),
        sql: None,
        error: None,
    }
}

/// Compare table data between connections
#[tauri::command]
pub async fn table_compare(
    _request: CompareTablesRequest,
) -> SchemaResponse {
    // Placeholder - would need database manager access for data comparison
    SchemaResponse {
        success: true,
        snapshots: None,
        snapshot: None,
        comparison: Some(serde_json::json!({
            "matches": 0,
            "differences": 0,
            "onlyInSource": 0,
            "onlyInTarget": 0,
            "rows": []
        })),
        sql: None,
        error: None,
    }
}

/// Generate migration SQL from comparison
#[tauri::command]
pub async fn migration_generate_sql(
    _request: GenerateMigrationSqlRequest,
) -> SchemaResponse {
    // Placeholder - would generate actual SQL based on comparison
    let sql = "-- Migration SQL\n-- Generated from schema comparison\n".to_string();
    
    SchemaResponse {
        success: true,
        snapshots: None,
        snapshot: None,
        comparison: None,
        sql: Some(sql),
        error: None,
    }
}

/// Generate sync SQL for data differences
#[tauri::command]
pub async fn migration_generate_sync_sql(
    _request: GenerateMigrationSqlRequest,
) -> SchemaResponse {
    // Placeholder - would generate sync SQL based on data comparison
    let sql = "-- Sync SQL\n-- Generated from data comparison\n".to_string();
    
    SchemaResponse {
        success: true,
        snapshots: None,
        snapshot: None,
        comparison: None,
        sql: Some(sql),
        error: None,
    }
}

File: apps/electron/src/main/services/ipc/schema.ts
import type {
  CompareConnectionsRequest,
  CompareConnectionToSnapshotRequest,
  CompareSnapshotsRequest,
  CompareTablesRequest,
  DeleteSchemaSnapshotRequest,
  GenerateMigrationSQLRequest,
  GenerateSyncSQLRequest,
  GetSchemaSnapshotRequest,
  SaveSchemaSnapshotRequest,
  SchemaSnapshot,
} from '@shared/types';
import { IPC_CHANNELS } from '@shared/types';
import { ipcMain } from 'electron';
import { dataDiffService } from '../data-diff';
import { dataDiffSyncGeneratorService } from '../data-diff-sync-generator';
import { databaseService } from '../database';
import { migrationGeneratorService } from '../migration-generator';
import { schemaComparisonService } from '../schema-comparison';
import {
  deleteSchemaSnapshot,
  getSchemaSnapshot,
  getSchemaSnapshots,
  saveSchemaSnapshot,
} from '../store';
import { createHandler } from './utils';

export function setupSchemaHandlers(): void {
  // Schema: Get Snapshots
  ipcMain.handle(
    IPC_CHANNELS.SCHEMA_GET_SNAPSHOTS,
    createHandler(async () => {
      const snapshots = getSchemaSnapshots();
      return { success: true, snapshots };
    })
  );

  // Schema: Get Snapshot
  ipcMain.handle(
    IPC_CHANNELS.SCHEMA_GET_SNAPSHOT,
    createHandler(async (request: GetSchemaSnapshotRequest) => {
      const snapshot = getSchemaSnapshot(request.snapshotId);
      return { success: true, snapshot };
    })
  );

  // Schema: Save Snapshot
  ipcMain.handle(
    IPC_CHANNELS.SCHEMA_SAVE_SNAPSHOT,
    createHandler(async (request: SaveSchemaSnapshotRequest) => {
      const snapshotData: SchemaSnapshot = {
        id: crypto.randomUUID(),
        name: request.name || 'Unnamed Snapshot',
        schemas: request.schema || [],
        connectionPath: request.connectionPath || '',
        description: request.description,
        createdAt: new Date().toISOString(),
      };
      saveSchemaSnapshot(snapshotData);
      return { success: true, snapshot: snapshotData };
    })
  );

  // Schema: Delete Snapshot
  ipcMain.handle(
    IPC_CHANNELS.SCHEMA_DELETE_SNAPSHOT,
    createHandler(async (request: DeleteSchemaSnapshotRequest) => {
      deleteSchemaSnapshot(request.snapshotId);
      return { success: true };
    })
  );

  // Schema: Compare Snapshots
  ipcMain.handle(
    IPC_CHANNELS.SCHEMA_COMPARE_SNAPSHOTS,
    createHandler(async (request: CompareSnapshotsRequest) => {
      const snapshotId1 = request.snapshotId1 || request.sourceSnapshotId || '';
      const snapshotId2 = request.snapshotId2 || request.targetSnapshotId || '';
      const snapshot1 = getSchemaSnapshot(snapshotId1);
      const snapshot2 = getSchemaSnapshot(snapshotId2);

      if (!snapshot1 || !snapshot2) {
        throw new Error('One or both snapshots not found');
      }

      const comparison = schemaComparisonService.compareSchemas(
        snapshot1.schemas || snapshot1.schema || [],
        snapshot2.schemas || snapshot2.schema || [],
        snapshot1.id,
        snapshot1.name,
        'snapshot',
        snapshot2.id,
        snapshot2.name,
        'snapshot'
      );
      return { success: true, comparison };
    })
  );

  // Schema: Compare Connections
  ipcMain.handle(
    IPC_CHANNELS.SCHEMA_COMPARE_CONNECTIONS,
    createHandler(async (request: CompareConnectionsRequest) => {
      const connectionId1 =
        request.connectionId1 || request.sourceConnectionId || '';
      const connectionId2 =
        request.connectionId2 || request.targetConnectionId || '';
      const schema1 = await databaseService.getSchema(connectionId1);
      const schema2 = await databaseService.getSchema(connectionId2);

      if (!schema1.success || !schema2.success) {
        throw new Error('Failed to fetch schemas');
      }

      const comparison = schemaComparisonService.compareSchemas(
        schema1.schemas || [],
        schema2.schemas || [],
        connectionId1,
        connectionId1,
        'connection',
        connectionId2,
        connectionId2,
        'connection'
      );
      return { success: true, comparison };
    })
  );

  // Schema: Compare Connection to Snapshot
  ipcMain.handle(
    IPC_CHANNELS.SCHEMA_COMPARE_CONNECTION_TO_SNAPSHOT,
    createHandler(async (request: CompareConnectionToSnapshotRequest) => {
      const connectionId = request.connectionId || '';
      const snapshotId = request.snapshotId || '';
      const liveSchema = await databaseService.getSchema(connectionId);
      const snapshot = getSchemaSnapshot(snapshotId);

      if (!liveSchema.success || !snapshot) {
        throw new Error('Failed to fetch schema or snapshot');
      }

      const comparison = schemaComparisonService.compareSchemas(
        liveSchema.schemas || [],
        snapshot.schemas || snapshot.schema || [],
        connectionId,
        connectionId,
        'connection',
        snapshot.id,
        snapshot.name,
        'snapshot'
      );
      return { success: true, comparison };
    })
  );

  // Table: Compare Tables
  ipcMain.handle(
    IPC_CHANNELS.TABLE_COMPARE,
    createHandler(async (request: CompareTablesRequest) => {
      const connectionId1 =
        request.connectionId1 || request.sourceConnectionId || '';
      const connectionId2 =
        request.connectionId2 || request.targetConnectionId || '';
      const table1 = request.table1 || request.sourceTable || '';
      const table2 = request.table2 || request.targetTable || '';
      const schema1 = request.schema1 || request.sourceSchema || '';
      const schema2 = request.schema2 || request.targetSchema || '';

      const comparison = await dataDiffService.compareTableData(
        connectionId1,
        table1,
        schema1,
        connectionId2,
        table2,
        schema2,
        request.primaryKeys || []
      );
      return { success: true, comparison };
    })
  );

  // Migration: Generate SQL
  ipcMain.handle(
    IPC_CHANNELS.MIGRATION_GENERATE_SQL,
    createHandler(async (request: GenerateMigrationSQLRequest) => {
      const result = migrationGeneratorService.generateMigrationSQL(request);
      return result;
    })
  );

  // Migration: Generate Sync SQL
  ipcMain.handle(
    IPC_CHANNELS.MIGRATION_GENERATE_SYNC_SQL,
    createHandler(async (request: GenerateSyncSQLRequest) => {
      const result = dataDiffSyncGeneratorService.generateSyncSQL(request);
      return result;
    })
  );
}
